### PR TITLE
20231101 start working on bulk update

### PR DESCRIPTION
**Pull request by uncomfyhalomacro**

### CODE REVIEW

File: bulk-updater/src/cli.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+use crate::operations;
+
+#[derive(Parser, Debug)]
+#[command(
+    author,
+    name = "bulk_updater",
+    about = "OBS Service Cargo Bulk Updater for Rust projects.",
+    after_long_help = "Set verbosity and tracing through `RUST_LOG` environmental variable e.g. `RUST_LOG=trace`
+
+Bugs can be reported on GitHub: https://github.com/uncomfyhalomacro/obs-service-cargo_vendor-rs/issues",
```
Comment (Firstyear) [inline bulk-updater/src/cli.rs:23]: Wrong github link?

File: bulk-updater/src/cli.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    max_term_width = 120
+)]
+pub struct BulkUpdaterOpts {
+    #[arg(
+        long,
+        default_value = "home:firstyear:branches",
+        help = "Custom basepath to run bulk updater."
```
Comment (Firstyear) [inline bulk-updater/src/cli.rs:30]: Run against this obs project

File: bulk-updater/src/operations.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+pub fn checkout_or_update(pkgname: &str, basepath: &Path) -> io::Result<PathBuf> {
+    let pkgpath = std::path::PathBuf::from(format!("{}:{}", basepath.to_string_lossy(), pkgname));
+    if pkgpath.exists() {
+        tracing::info!("osc revert {}", pkgpath.to_string_lossy());
+        let arguments = vec!["revert", "."];
+        let out = osc_command(&pkgpath, &arguments)?;
+        if out.status.success() {
+            let command_output = String::from_utf8_lossy(&out.stdout);
+            tracing::info!("✅ -- osc revert operation passed");
```
Comment (Firstyear) [inline bulk-updater/src/operations.rs:66]: minor, success not passed.

File: bulk-updater/src/operations.rs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        let out = osc_command(&pkgpath, &arguments)?;
+        if out.status.success() {
+            let command_output = String::from_utf8_lossy(&out.stdout);
+            tracing::info!("✅ -- osc clean operation passed");
+            tracing::info!("stdout -- {}", command_output);
+        } else {
+            let command_output = String::from_utf8_lossy(&out.stderr);
+            tracing::info!("🚨 -- osc clean operation failed");
+            tracing::info!("stderr -- {}", command_output);
+        };
```
Comment (Firstyear) [inline bulk-updater/src/operations.rs:83]: Let the code breath. Have spaces between "logical blocks". It's okay. That whitespace won't hurt, and helps people read and delineate things.

File: bulk-updater/src/operations.rs [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        if out.status.success() {
+            let command_output = String::from_utf8_lossy(&out.stdout);
+            tracing::info!("✅ -- osc ci operation passed");
+            tracing::info!("stdout -- {}", command_output);
+        } else {
+            let command_output = String::from_utf8_lossy(&out.stderr);
+            tracing::info!("🚨 -- osc ci operation failed");
+            tracing::info!("stderr -- {}", command_output);
+        };
+        let arguments = vec!["sr", "-m", message];
```
Comment (Firstyear) [inline bulk-updater/src/operations.rs:136]: maybe we need an extra option for "opts.findout" which does the submit ontop of yolo. That way we can do a yolo commit, but then manually sr, or we can "findout" and just go ham.

File: bulk-updater/src/operations.rs [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            let command_output = String::from_utf8_lossy(&out.stdout);
+            tracing::info!("✅ -- osc sr operation passed");
+            tracing::info!("stdout -- {}", command_output);
+        } else {
+            let command_output = String::from_utf8_lossy(&out.stderr);
+            tracing::info!("🚨 -- osc sr operation failed");
+            tracing::info!("stderr -- {}", command_output);
+        };
+    } else {
+        tracing::info!("🤌 You must manually run the following in {pkgpath_str}:\n`osc vc -m {message}`\n`osc ci -m {message}`\n`osc sr -m {message}`");
```
Comment (Firstyear) [inline bulk-updater/src/operations.rs:148]: Amazing choice of emoji.

File: bulk-updater/src/cli.rs (hunk 1 above)
Comment (uncomfyhalomacro) [reply bulk-updater/src/cli.rs:23]: truth

File: bulk-updater/src/operations.rs (hunk 5 above)
Comment (uncomfyhalomacro) [reply bulk-updater/src/operations.rs:136]: so findout is a param if we want to manually sr or not?

File: bulk-updater/src/operations.rs (hunk 6 above)
Comment (uncomfyhalomacro) [reply bulk-updater/src/operations.rs:148]: 😔

File: bulk-updater/src/operations.rs (hunk 4 above)
Comment (uncomfyhalomacro) [reply bulk-updater/src/operations.rs:83]: i did somethings but cargo fmt be like 🔫

File: bulk-updater/src/cli.rs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            tracing::info!("🥳 Updated {}", pkgname.to_string_lossy());
+        }
+        for pkgpath in pkgpaths.iter() {
+            tracing::info!(
+                "🔼 Attempting to update in progress at {}",
+                pkgpath.to_string_lossy()
+            );
+            let updated_pkgpath = operations::attempt_update(pkgpath, self.color)?;
+            tracing::info!("✅ Updated package at {}", pkgpath.to_string_lossy());
+            tracing::info!(
```
Comment (Firstyear) [inline bulk-updater/src/cli.rs:84]: We might want to do the submits as a big loop at the end rather that between the updates.

File: bulk-updater/src/cli.rs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                "⏫ Checkout or update in progress for {}",
+                pkgname.to_string_lossy()
+            );
+            pkgpaths.push(operations::checkout_or_update(
+                &pkgname.to_string_lossy(),
+                &self.basepath,
+            )?);
+            tracing::info!("🥳 Updated {}", pkgname.to_string_lossy());
+        }
+        for pkgpath in pkgpaths.iter() {
```
Comment (Firstyear) [inline bulk-updater/src/cli.rs:77]: For bonus marks, make this par_iter with rayon since these can all update in parallel.

**Review by Firstyear — COMMENTED** [review]
Just some minor comments now, otherwise the rayon comments aren't super important, you can do that anytime.

File: bulk-updater/src/cli.rs (hunk 7 above)
Comment (uncomfyhalomacro) [reply bulk-updater/src/cli.rs:84]: sure :D

File: bulk-updater/src/cli.rs (hunk 8 above)
Comment (uncomfyhalomacro) [reply bulk-updater/src/cli.rs:77]: yeah i could try using rayon. it's smaller than tokio

**Review by Firstyear — APPROVED** [review]
This looks awesome, I'll have to give it a try soon!

**Pull request merged**